Tasks/Npm/npmtask.ts

```diff
@@ -1,27 +1,189 @@
+    try{
+        var execOptions : trm.IExecOptions = {};
+        execOptions.env = process.env || getBuildCredProviderEnv();
+        var code : number = await npmAuthRunner.exec(execOptions);
+        tl.debug('Authentication succeeded with code: ' + code);
+        return Q(code);
+    } catch (err) {
+        tl.debug(tl.loc('NpmAuthFailed', err.message));
+        throw err;
+    }
+}
+
+async function tryRunNpmConfigAsync(npmConfigRunner: trm.ToolRunner, execOptions : trm.IExecOptions) {
+    try {
+        var code = await npmConfigRunner.exec(execOptions);
+    } catch (err) {
+        // do not throw on this failure.
+        tl.warning(tl.loc('NpmConfigFailed', err.Message));
+    }
+}
+
+async function runNpmCommandAsync(npmCommandRunner: trm.ToolRunner, execOptions : trm.IExecOptions) : Promise<number> {
+    try{
+        var code: number = await npmCommandRunner.exec(execOptions);
+        tl.debug('Npm command succeeded with code: ' + code);
+        return Q(code);
+    } catch (err) {
+        tl.debug(tl.loc('NpmFailed', err.message));
+        throw err;
+    }
 }
-npm.arg(command);
 
-npm.argString(tl.getInput('arguments', false));
+function shouldUseDeprecatedTask() : boolean {
+    return tl.getVariable('USE_DEPRECATED_TASK_VERSION') != undefined && tl.getVariable('USE_DEPRECATED_TASK_VERSION').toLowerCase() === 'true';
+}
+
+function getNpmAuthHelperRunner(npmrcPath: string, tempUserNpmrcPath: string, includeDebugLogs: boolean): trm.ToolRunner {
+    var npmAuthHelper = tl.createToolRunner(path.join(__dirname, 'Npm/vsts-npm-auth/bin/vsts-npm-auth.exe'));
+    var verbosityString = includeDebugLogs ? 'Detailed' : 'Normal';
+    npmAuthHelper.argString(`-NonInteractive -Verbosity ${verbosityString} -Config "${npmrcPath}" -TargetConfig "${tempUserNpmrcPath}"`);
+    return npmAuthHelper;
+}
+
+function getNpmConfigRunner(includeDebugLogs: boolean): trm.ToolRunner {
+    var npmConfig = tl.createToolRunner(tl.which('npm', true));
+    npmConfig.argString('config list');
+    if(includeDebugLogs) {
+        npmConfig.arg('-l');
+    }
+    return npmConfig;
+}
+
+function getTempNpmrcPath() : string {
+    var tempNpmrcDir
+        = tl.getVariable('Agent.BuildDirectory')
+        || tl.getVariable('Agent.ReleaseDirectory')
+        || process.cwd();
+    tempNpmrcDir = path.join(tempNpmrcDir, 'npm');
+    tl.mkdirP(tempNpmrcDir);
+    var tempUserNpmrcPath: string = path.join(tempNpmrcDir, 'auth.' + tl.getVariable('build.buildId') + '.npmrc');
+    return tempUserNpmrcPath;
+}
+
+function cleanUpTempNpmrcPath(tempUserNpmrcPath: string) {
+    tl.debug('cleaning up...')
+    if(tl.exist(tempUserNpmrcPath)) {
+        tl.debug('deleting temporary npmrc...');
+        tl.rmRF(tempUserNpmrcPath, /* continueOnError */ false);
+    }
+}
+
+function getBuildCredProviderEnv() : EnvironmentDictionary {
+
+	var env : EnvironmentDictionary = {};
+    let credProviderPath : string = path.join(__dirname, 'NuGet/CredentialProvider');;
```

Comment:
NuGet?
Also - could we use "which" to get the path to CredentialProvider.Vss.exe, then find it's root dir? Again would allow easier L0 testing and would move path resolution out of this portion of the code